ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/OptimizedMerkleTrieLoader.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Hyperledger Besu contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.bonsai;
+
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.trie.MerklePatriciaTrie;
+import org.hyperledger.besu.ethereum.trie.StoredMerklePatriciaTrie;
+import org.hyperledger.besu.metrics.BesuMetricCategory;
+import org.hyperledger.besu.metrics.ObservableMetricsSystem;
+import org.hyperledger.besu.metrics.prometheus.PrometheusMetricsSystem;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.prometheus.client.guava.cache.CacheMetricsCollector;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class OptimizedMerkleTrieLoader {
+
+  private static final int ACCOUNT_CACHE_SIZE = 100_000;
+  private static final int STORAGE_CACHE_SIZE = 200_000;
+  private final Cache<Bytes, Bytes> accountsNodes;
+  private final Cache<Bytes, Bytes> storageNodes;
+
+  public OptimizedMerkleTrieLoader(final ObservableMetricsSystem metricsSystem) {
+    accountsNodes = CacheBuilder.newBuilder().recordStats().maximumSize(ACCOUNT_CACHE_SIZE).build();
+    storageNodes = CacheBuilder.newBuilder().recordStats().maximumSize(STORAGE_CACHE_SIZE).build();
+
+    CacheMetricsCollector cacheMetrics = new CacheMetricsCollector();
+    cacheMetrics.addCache("accountsNodes", accountsNodes);
+    cacheMetrics.addCache("storageNodes", storageNodes);
+    if (metricsSystem instanceof PrometheusMetricsSystem)
+      ((PrometheusMetricsSystem) metricsSystem)
+          .addCollector(BesuMetricCategory.BLOCKCHAIN, () -> cacheMetrics);
+  }
+
+  public void preLoadAccount(
+      final BonsaiWorldStateKeyValueStorage worldStateStorage,
+      final Hash worldStateRootHash,
+      final Address account) {
+    CompletableFuture.runAsync(
```

Comment:
It feels to me weird that nobody handles the future when it is done. Maybe you can have the Optional<Node> that you were preloading to be a return value of the future and when the future completes, update your accountsNodes. That would also prevent the need of modifying a field in the middle of a function. Functions in functional programming should preferably not have side effects, so this might help.

Reply:
Good point, the idea was to execute asynchronously as much work as possible to fill the cache. These completable futures doesn't return any result, but I totally agree that it should be more clean to control the execution of the completable future. 
This is a first version, and this can be improved in a future PR. The control of completable futures' executions will introduce a lot of changes that can be done in the future PR. We can for example stop all the futures when we start reading from the cache, this is not done in this current PR.